core/src/main/java/org/bitcoinj/utils/VersionTally.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2015 Ross Nicoll.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bitcoinj.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.bitcoinj.core.NetworkParameters;
+
+/**
+ * Caching counter for the block versions within a moving window. Thread
+ * safe.
+ *
+ * @see org.bitcoinj.core.NetworkParameters#getMajorityWindow()
+ */
+public class VersionTally {
+    /**
+     * Current count for each version number. In theory could use an array for
+     * this, however means any out of sequence version numbers would require a
+     * much larger array.
+     */
+    private final Map<Long, Integer> counts = new HashMap<Long, Integer>();
+
+    /**
+     * Cache of version numbers.
+     */
+    private final long[] versionWindow;
+
+    /**
+     * Offset within the version window at which the next version will be
+     * written.
+     */
+    private int versionWriteHead = 0;
+
+    /**
+     * Number of versions written into the tally. Until this matches the length
+     * of the version window, we do not have sufficient data to return values.
+     */
+    private int versionsStored = 0;
+
+    private final Object lock = new Object();
```

Comment:
Good point, although in hindsight the class cannot be thread safe (the order of versions being added is important), so ripping the synchronization out anyway!